### PR TITLE
Modernize Kernel.Typespec by reducing usage of :lists functions

### PR DESCRIPTION
**Pull request by whatyouhide**

### CODE REVIEW

File: lib/elixir/lib/kernel/typespec.ex [hunk 1]
```diff
@@ -174,11 +174,9 @@ defmodule Kernel.Typespec do
   @spec defines_spec?(module, atom, arity) :: boolean
   def defines_spec?(module, name, arity)
       when is_atom(module) and is_atom(name) and arity in 0..255 do
-    finder = fn {_kind, expr, _caller} ->
+    Enum.any?(Module.get_attribute(module, :spec), fn {_kind, expr, _caller} ->
       spec_to_signature(expr) == {name, arity}
-    end
-
-    :lists.any(finder, Module.get_attribute(module, :spec))
+    end)
```
Comment (whatyouhide) [inline lib/elixir/lib/kernel/typespec.ex]: I was able to do this here and in `defines_callback?/3` below but I ran into bootstrapping problems doing this in https://github.com/elixir-lang/elixir/pull/7437/files#diff-6de4b356bc9d4e3d07f1e5731fbe7d23L167 (`defines_type?/3`), where it complained that `Enum.any?/2` was undefined 🤔

File: lib/elixir/lib/kernel/typespec.ex [hunk 2]
```diff
@@ -577,21 +571,18 @@ defmodule Kernel.Typespec do
...
           [{:var, line, name}, typespec(type, vars, caller)]
         ]
 
-        type = {:type, line, :constraint, constraint}
-        [type | acc]
-    end
-
-    :lists.reverse(:lists.foldl(foldl_fun, [], guard))
+        [{:type, line, :constraint, constraint}]
+    end)
```
Comment (whatyouhide) [inline lib/elixir/lib/kernel/typespec.ex]: This is really nice, was able to drop the reversing as well!

File: lib/elixir/lib/kernel/typespec.ex [hunk 3]
```diff
@@ -1154,10 +1142,13 @@ defmodule Kernel.Typespec do
   end
 
   defp typespec(list, vars, caller) when is_list(list) do
-    [h | t] = :lists.reverse(list)
```
Comment (eksperimental) [inline lib/elixir/lib/kernel/typespec.ex]: I think we should avoid single letter vars

File: lib/elixir/lib/kernel/typespec.ex [hunk 4]
```diff
@@ -1154,10 +1142,13 @@ defmodule Kernel.Typespec do
...
+    [h | t] = Enum.reverse(list)
+
+    union =
+      Enum.reduce(t, validate_kw(h, list, caller), fn x, acc ->
```
Comment (eksperimental) [inline lib/elixir/lib/kernel/typespec.ex]: Same here

File: lib/elixir/lib/kernel/typespec.ex (hunk 1 above)
Comment (josevalim) [reply lib/elixir/lib/kernel/typespec.ex]: defines_type? is used by Protocols and we need protocols to compile Enum. So I guess we can't do it for defines_type? although that's probably fine.

Comment (whatyouhide) [reply lib/elixir/lib/kernel/typespec.ex]: Ah I see, cool!

**Pull request merged**